fraud/service.go

```diff
@@ -115,30 +147,190 @@ func (f *service) processIncoming(
 		"hash", hex.EncodeToString(extHeader.DAH.Hash()),
 		"from", from.String(),
 	)
-	msg.ValidatorData = proof
+	log.Warn("Shutting down services...")
+	return pubsub.ValidationAccept
+}
+
+func (f *service) processLocal(
+	ctx context.Context,
+	proofType ProofType,
+	hash string,
+	rawData []byte,
+) pubsub.ValidationResult {
+	f.storesLk.RLock()
+	storage := f.stores[proofType]
+	f.storesLk.RUnlock()
+	data, err := getByHash(ctx, storage, hash)
+	if err != nil {
+		// TODO @vgonkivs: should we panic here???
+		panic(fmt.Sprintf("unexpected error while fetching a proof from the storage: %s", err.Error()))
+	}
+
+	if bytes.Compare(data, rawData) == 0 {
+		log.Warnw("received fraud proof",
+			"proofType", proofType,
+			"headerHash", hash,
+		)
+		return pubsub.ValidationAccept
+	}
+	return pubsub.ValidationReject
+}
+
+func (f *service) Get(ctx context.Context, proofType ProofType) ([]Proof, error) {
 	f.storesLk.Lock()
 	store, ok := f.stores[proofType]
 	if !ok {
 		store = initStore(proofType, f.ds)
 		f.stores[proofType] = store
 	}
 	f.storesLk.Unlock()
-	err = put(ctx, store, string(proof.HeaderHash()), msg.Data)
-	if err != nil {
-		log.Error(err)
-	}
-	log.Warn("Shutting down services...")
-	return pubsub.ValidationAccept
+
+	return getAll(ctx, store, proofType)
 }
 
-func (f *service) Get(ctx context.Context, proofType ProofType) ([]Proof, error) {
+func (f *service) put(ctx context.Context, proofType ProofType, hash string, data []byte) error {
 	f.storesLk.Lock()
 	store, ok := f.stores[proofType]
 	if !ok {
 		store = initStore(proofType, f.ds)
 		f.stores[proofType] = store
 	}
 	f.storesLk.Unlock()
+	return put(ctx, store, hash, data)
+}
 
-	return getAll(ctx, store, proofType)
+func (f *service) handleFraudMessageRequest(stream network.Stream) {
+	msg := new(pb.FraudMessage)
+	_, err := serde.Read(stream, msg)
+	if err != nil {
+		stream.Reset() //nolint:errcheck
+		log.Warn(err)
```

Comment:
Can we add more descriptive and unique messaging here with the error itself? 
Sometimes searching in the code base through those unique hand-written messages helps a lot in tracing what happened

Reply:
If the above is ok, then doing it in a separate PR would be awesome!